# Data Analysis with Pandas

### Import and use Pandas

In [1]:
import pandas as pd

# creating new series
number_series = pd.Series([1, 2, 3, 4])
print(number_series)

# create new Data Frame
user_data_dict = {
    "Name": ["Rishav", "Krishna"],
    "Age": [21, 23],
    "Address": ["Kathmandu", "Lalitpur"]
}
user_data = pd.DataFrame(user_data_dict)
user_data

0    1
1    2
2    3
3    4
dtype: int64


Name  Age    Address
0   Rishav   21  Kathmandu
1  Krishna   23   Lalitpur

### Reading data from CSV, Excel and JSON file
#### Read data from csv

In [3]:
import pandas as pd

student_csv_data = pd.read_csv("data/student.csv")
student_csv_data.head()

id        name  class  mark  gender
0   1    John Deo   Four  75.0  female
1   2    Max Ruin  Three  85.0    male
2   3      Arnold  Three  55.0    male
3   4  Krish Star   Four  60.0  female
4   5   John Mike   Four  60.0  female

### Save Pandas Dataframe to csv, excel and JSON file

In [4]:
import pandas as pd

# Creating new data frame
user_data = pd.DataFrame({
    "Name": ["Rishav", "Krishna"],
    "Age": [21, 23],
    "Address": ["Kathmandu", "Lalitpur"]
})

user_data.to_csv('./data/user_data.csv')
user_data.to_excel('data/user_data.xlsx')

### Basic Data Exploration
#### Data Inspection

In [13]:
import pandas as pd

students = pd.read_csv('data/student.csv')
print(students.shape)

print(students.head()) # print top n=5 rows
print(students.info()) # print information regarding the dataframe
students.describe()

(35, 5)
   id        name  class  mark  gender
0   1    John Deo   Four  75.0  female
1   2    Max Ruin  Three  85.0    male
2   3      Arnold  Three  55.0    male
3   4  Krish Star   Four  60.0  female
4   5   John Mike   Four  60.0  female
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35 entries, 0 to 34
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   id      35 non-null     int64  
 1   name    35 non-null     object 
 2   class   35 non-null     object 
 3   mark    33 non-null     float64
 4   gender  34 non-null     object 
dtypes: float64(1), int64(1), object(3)
memory usage: 1.5+ KB
None


id       mark
count  35.000000  33.000000
mean   18.000000  74.121212
std    10.246951  16.714061
min     1.000000  18.000000
25%     9.500000  60.000000
50%    18.000000  78.000000
75%    26.500000  88.000000
max    35.000000  96.000000

#### Single Value Access (.at[] and .iat[])

- .at[] is used for accessing a single scalar value by label
- .iat[] is for accessing by index position

In [ ]:
# using .at[] to access single value by label
print(students.at[1, 'name']) # Value in row 2, column "name"

# Using .iat[] to access single value by position
print(students.iat[5, 1]) # Value in row 6, column 1

Max Ruin
Alex John


### Filtering

- Logical operators (>, <, ==, !=)
- Multiple logical operators (& (AND) and | (OR))
- Str ancessor, Isin, contains, between, startWith
- Tilde (~)
- Query
- Nlargest and nsmallest
- Loc and iloc

#### Logical Operators

In [8]:
# student with marks less than 30
students_marks_less_than_40 = students[students.mark<30]
print(students_marks_less_than_40.head())

boys = students[students.gender=='male']

boys.head()

    id   name class  mark gender
18  19  Tinny  Nine  18.0   male


id         name  class  mark gender
1   2     Max Ruin  Three  85.0   male
2   3       Arnold  Three  55.0   male
5   6    Alex John   Four  55.0   male
6   7  My John Rob  Fifth  78.0   male
7   8       Asruid   Five  85.0   male

#### str, lsin, contains, between, startWith

In [12]:
# students whose name starts with 'J'
students_with_name_j = students[students.name.str.startswith('J')]
students_with_name_j.head()

# students that contain name jimmy
students_with_name_gimmy = students[students.name.str.contains('Gimmy')]
print(students_with_name_gimmy)

# students with marks between 40 and 50
students_with_mark_between = students[students.mark.between(40,50)]
students_with_mark_between.head()

# students of class Nine or Ten
class_nine_or_ten = students[students['class'].isin(('Nine', 'Ten'))]
class_nine_or_ten

    id   name class  mark gender
15  16  Gimmy  Four  88.0   male


id    name class  mark  gender
18  19   Tinny  Nine  18.0    male
19  20  Jackly  Nine  65.0  female

#### Tilde(~)

In [13]:
# Filter students who are NOT in class 'Nine'
not_class_nine = students[~students['class'].isin(['Nine'])]
not_class_nine.head()

id        name  class  mark  gender
0   1    John Deo   Four  75.0  female
1   2    Max Ruin  Three  85.0    male
2   3      Arnold  Three  55.0    male
3   4  Krish Star   Four  60.0  female
4   5   John Mike   Four  60.0  female

#### Query

In [14]:
# Filter students with marks above 85 and gender is 'male'
query_filter = students.query("mark > 85 and gender == 'male'")
query_filter.head()

id      name  class  mark gender
14  15  Tade Row   Four  88.0   male
15  16     Gimmy   Four  88.0   male
24  25  Giff Tow  Seven  88.0   male

#### nlargest and nsmallest

In [4]:
# Get top 3 students with the highest marks
top_3_marks = students.nlargest(3, 'mark')
print(top_3_marks)
# Get 2 students with the lowest marks
bottom_2_marks = students.nsmallest(2, 'mark')
bottom_2_marks

    id       name  class  mark  gender
32  33  Kenn Rein    Six  96.0  female
11  12      Recky    Six  94.0  female
31  32  Binn Rott  Seven  90.0  female


id   name class  mark gender
18  19  Tinny  Nine  18.0   male
16  17  Tumyu   Six  54.0   male

### Data Transformation and Mapping
- **apply:** allows us to apply a function to each element, row, or column in a DataFrame or Series
- **map:** is used to map values in a Series according to a dictionary or another Series
- **replace:** replace allows for replacing specific values in the DataFrame with new value
- **astype:** this is used to convert data type
- **pipe:** it allows chaining and using complex functions that operate on the entire DataFrame. This can be used for complex transformation operation

#### apply

In [14]:
# triple the marks of the student using apply and lambda function
students['triple_mark'] = students['mark'].apply(lambda x: x * 3)
students.head()

id        name  class  mark  gender  triple_mark
0   1    John Deo   Four  75.0  female        225.0
1   2    Max Ruin  Three  85.0    male        255.0
2   3      Arnold  Three  55.0    male        165.0
3   4  Krish Star   Four  60.0  female        180.0
4   5   John Mike   Four  60.0  female        180.0

#### map

In [15]:
# replacing the 'male' with 1 and 'female' with 0
gender_map = {'male': 1, 'female': 0}
students['gender_code'] = students['gender'].map(gender_map)
students.head()

id        name  class  mark  gender  triple_mark  gender_code
0   1    John Deo   Four  75.0  female        225.0          0.0
1   2    Max Ruin  Three  85.0    male        255.0          1.0
2   3      Arnold  Three  55.0    male        165.0          1.0
3   4  Krish Star   Four  60.0  female        180.0          0.0
4   5   John Mike   Four  60.0  female        180.0          0.0

#### replace

In [16]:
# replacing the words to numbers in class column
students['class'] = students['class'].replace({
    "Four": 4,
    "Ten": 10,
    "Eight": 8
})

students.head()

id        name  class  mark  gender  triple_mark  gender_code
0   1    John Deo      4  75.0  female        225.0          0.0
1   2    Max Ruin  Three  85.0    male        255.0          1.0
2   3      Arnold  Three  55.0    male        165.0          1.0
3   4  Krish Star      4  60.0  female        180.0          0.0
4   5   John Mike      4  60.0  female        180.0          0.0

#### astype

In [17]:
# Converting integer to float
students['mark'] = students['mark'].astype(float)
students.head()

id        name  class  mark  gender  triple_mark  gender_code
0   1    John Deo      4  75.0  female        225.0          0.0
1   2    Max Ruin  Three  85.0    male        255.0          1.0
2   3      Arnold  Three  55.0    male        165.0          1.0
3   4  Krish Star      4  60.0  female        180.0          0.0
4   5   John Mike      4  60.0  female        180.0          0.0

#### pipe

In [18]:
# creating a function that doubles the marks and pipes it to the pandas dataframe
def double_marks(df):
    df["mark"] = df["mark"] * 2
    return df

students = students.pipe(double_marks)
students.head()

id        name  class   mark  gender  triple_mark  gender_code
0   1    John Deo      4  150.0  female        225.0          0.0
1   2    Max Ruin  Three  170.0    male        255.0          1.0
2   3      Arnold  Three  110.0    male        165.0          1.0
3   4  Krish Star      4  120.0  female        180.0          0.0
4   5   John Mike      4  120.0  female        180.0          0.0

### Data Cleaning and Manipulation

- Handling Missing Data
- Handling Duplicates
- One-Hot-Encoding
- Normalization

#### Handling Missing Data
- df.dropna() removes dropes values containing missing values
    - dropna(axis=0) removes rows containing at least one missing value
    - dropna(axis = 1) removes columns containig atleast one missing value
- df.fillna() replaces the missing values
    - df.fillna(specific values)
    - df.fillna(value, method="ffill") replaces the missing values with the next value
    - df.fillna({'height': 150, 'weight': 60}) replaces the missing values of column height with 150 and column weight with 60.
    - fill with average, interpolation: fill the missing value using average, or using interpolation

In [22]:
from IPython.display import display

# axis, 0=row, 1=column
weight_height = pd.read_csv("./data/weight-height.csv")

display(weight_height.head())

Gender     Height      Weight Country
0   Male  73.847017  241.893563   India
1   Male  68.781904  162.310473   Nepal
2   Male  74.110105  212.740856   India
3   Male  71.730978  220.042470   Nepal
4   Male  69.881796  206.349801   Nepal

In [21]:
# (axis=0 default) drops rows with NaN
weight_height.dropna(axis=0, how="any", inplace=True)
display(weight_height.head())

Gender     Height      Weight Country
0   Male  73.847017  241.893563   India
1   Male  68.781904  162.310473   Nepal
2   Male  74.110105  212.740856   India
3   Male  71.730978  220.042470   Nepal
4   Male  69.881796  206.349801   Nepal

In [ ]:
weight_height.fillna(method="ffill")
display(weight_height.head())
# weight_height.fillna(method="bfill") to fill with next row

/tmp/ipykernel_9622/1932057922.py:1: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  weight_height.fillna(method="ffill")


Gender     Height      Weight Country
0   Male  73.847017  241.893563   India
1   Male  68.781904  162.310473   Nepal
2   Male  74.110105  212.740856   India
3   Male  71.730978  220.042470   Nepal
4   Male  69.881796  206.349801   Nepal

In [26]:
df_interpolated = weight_height.interpolate()
weight_height.fillna(df_interpolated)

/tmp/ipykernel_9622/1677365979.py:1: FutureWarning: DataFrame.interpolate with object dtype is deprecated and will raise in a future version. Call obj.infer_objects(copy=False) before interpolating instead.
  df_interpolated = weight_height.interpolate()


Gender     Height      Weight Country
0       Male  73.847017  241.893563   India
1       Male  68.781904  162.310473   Nepal
2       Male  74.110105  212.740856   India
3       Male  71.730978  220.042470   Nepal
4       Male  69.881796  206.349801   Nepal
...      ...        ...         ...     ...
9995  Female  66.172652  136.777454   India
9996  Female  67.067155  170.867906   Nepal
9997  Female  63.867992  128.475319   Nepal
9998  Female  69.034243  163.852461   Nepal
9999  Female  61.944246  113.649103   India

[10000 rows x 4 columns]

### Data normalization and Numpy